File: dotnet-internal/Services/SessionService.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using IdentityExample.Models;

namespace IdentityExample.Services;

/// <summary>
/// ECDH session management.
/// Step 1: Anonymous session init (30 min TTL).
/// Step 4: Authenticated session refresh with Bearer + X-Subject (1 hr TTL).
/// </summary>
public sealed class SessionService
{
    private readonly CryptoService _crypto;
    private readonly IHttpClientFactory _httpClientFactory;
    private readonly string _clientId;
    private readonly string _sidecarUrl;

    public SessionService(CryptoService crypto, IHttpClientFactory httpClientFactory,
                          IConfiguration config)
    {
        _crypto = crypto;
        _httpClientFactory = httpClientFactory;
        _clientId = config["Client:Id"] ?? "dev-client";
        _sidecarUrl = config["Sidecar:Url"] ?? "http://localhost:8141";
    }

    /// <summary>Step 1: Anonymous ECDH session initialization.</summary>
    public async Task<SessionContext> InitSessionAsync()
    {
        using var ecdh = _crypto.GenerateEcdhKeyPair();
        var pubKeyBytes = _crypto.ExportPublicKey(ecdh);
        var nonce = _crypto.GenerateNonce();
        var ts = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds().ToString();

        var body = JsonSerializer.Serialize(new SessionInitRequest(_crypto.ToBase64(pubKeyBytes)));

        using var client = _httpClientFactory.CreateClient("sidecar");
        using var request = new HttpRequestMessage(HttpMethod.Post, $"{_sidecarUrl}/api/v1/session/init");
        request.Content = new StringContent(body, Encoding.UTF8, "application/json");
        request.Headers.Add("X-Idempotency-Key", $"{ts}.{nonce}");
        request.Headers.Add("X-ClientId", _clientId);

        var response = await client.SendAsync(request);
        var responseBody = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
            throw new Exception($"Session init failed: HTTP {(int)response.StatusCode} — {responseBody}");

        var data = JsonSerializer.Deserialize<SessionInitResponse>(responseBody)!;
        return DeriveSession(ecdh, data, authenticated: false);
    }

    /// <summary>Step 4: Authenticated session refresh.</summary>
    public async Task<SessionContext> RefreshSessionAsync(
        string accessToken, string subject, SessionContext? oldSession)
    {
        using var ecdh = _crypto.GenerateEcdhKeyPair();
        var pubKeyBytes = _crypto.ExportPublicKey(ecdh);
        var nonce = _crypto.GenerateNonce();
        var ts = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds().ToString();

        var body = JsonSerializer.Serialize(new SessionInitRequest(_crypto.ToBase64(pubKeyBytes)));

        using var client = _httpClientFactory.CreateClient("sidecar");
        using var request = new HttpRequestMessage(HttpMethod.Post, $"{_sidecarUrl}/api/v1/session/init");
        request.Content = new StringContent(body, Encoding.UTF8, "application/json");
        request.Headers.Add("Authorization", $"Bearer {accessToken}");
        request.Headers.Add("X-Subject", subject);
        request.Headers.Add("X-Idempotency-Key", $"{ts}.{nonce}");
        request.Headers.Add("X-ClientId", _clientId);

        var response = await client.SendAsync(request);
        var responseBody = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
            throw new Exception($"Session refresh failed: HTTP {(int)response.StatusCode} — {responseBody}");

        // Zeroize old session key
        oldSession?.Zeroize();

        var data = JsonSerializer.Deserialize<SessionInitResponse>(responseBody)!;
        return DeriveSession(ecdh, data, authenticated: true);
    }

    private SessionContext DeriveSession(ECDiffieHellman ecdh, SessionInitResponse data, bool authenticated)
    {
        var serverPubBytes = _crypto.FromBase64(data.ServerPublicKey);
        var shared = _crypto.ComputeSharedSecret(ecdh, serverPubBytes);

        var salt = Encoding.UTF8.GetBytes(data.SessionId);
        var info = Encoding.UTF8.GetBytes($"SESSION|A256GCM|{_clientId}");
        var sessionKey = _crypto.Hkdf(shared, salt, info, 32);

        CryptographicOperations.ZeroMemory(shared);

        return new SessionContext(
            data.SessionId, sessionKey, $"session:{data.SessionId}",
            _clientId, authenticated, data.ExpiresInSec);
    }
}

File: flutter-external/lib/services/session_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../config.dart';
import '../models/api_models.dart';
import '../models/session_context.dart';
import 'crypto_service.dart';

/// ECDH session management.
/// Step 1: Anonymous session init (30 min TTL).
/// Step 4: Authenticated session refresh with Bearer + X-Subject (1 hr TTL).
class SessionService {
  final CryptoService _crypto;

  SessionService(this._crypto);

  /// Step 1: Anonymous ECDH session initialization.
  Future<SessionContext> initSession() async {
    final keyPair = await _crypto.generateEcdhKeyPair();
    final pubKeyBytes = await _crypto.exportPublicKey(keyPair);
    final nonce = _crypto.generateNonce();
    final ts = DateTime.now().millisecondsSinceEpoch.toString();

    final body = jsonEncode(
      SessionInitRequest(clientPublicKey: _crypto.toBase64(pubKeyBytes)).toJson(),
    );

    final response = await http.post(
      Uri.parse('${AppConfig.sidecarUrl}/api/v1/session/init'),
      headers: {
        'Content-Type': 'application/json',
        'X-Idempotency-Key': '$ts.$nonce',
        'X-ClientId': AppConfig.clientId,
      },
      body: body,
    );

    if (response.statusCode != 200) {
      throw Exception(
        'Session init failed: HTTP ${response.statusCode} — ${response.body}',
      );
    }

    final data = SessionInitResponse.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );

    return _deriveSession(keyPair, data, authenticated: false);
  }

  /// Step 4: Authenticated session refresh.
  Future<SessionContext> refreshSession(
    String accessToken,
    String subject,
    SessionContext? oldSession,
  ) async {
    final keyPair = await _crypto.generateEcdhKeyPair();
    final pubKeyBytes = await _crypto.exportPublicKey(keyPair);
    final nonce = _crypto.generateNonce();
    final ts = DateTime.now().millisecondsSinceEpoch.toString();

    final body = jsonEncode(
      SessionInitRequest(clientPublicKey: _crypto.toBase64(pubKeyBytes)).toJson(),
    );

    final response = await http.post(
      Uri.parse('${AppConfig.sidecarUrl}/api/v1/session/init'),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $accessToken',
        'X-Subject': subject,
        'X-Idempotency-Key': '$ts.$nonce',
        'X-ClientId': AppConfig.clientId,
      },
      body: body,
    );

    if (response.statusCode != 200) {
      throw Exception(
        'Session refresh failed: HTTP ${response.statusCode} — ${response.body}',
      );
    }

    // Zeroize old session key
    oldSession?.zeroize();

    final data = SessionInitResponse.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );

    return _deriveSession(keyPair, data, authenticated: true);
  }

  Future<SessionContext> _deriveSession(
    EcdhKeyPair keyPair,
    SessionInitResponse data, {
    required bool authenticated,
  }) async {
    final shared = await _crypto.computeSharedSecret(
      keyPair,
      data.serverPublicKey,
    );

    final sessionKey = await _crypto.deriveSessionKey(
      shared,
      data.sessionId,
      AppConfig.clientId,
    );

    return SessionContext(
      sessionId: data.sessionId,
      sessionKey: sessionKey,
      kid: 'session:${data.sessionId}',
      clientId: AppConfig.clientId,
      authenticated: authenticated,
      expiresInSec: data.expiresInSec,
    );
  }
}
